Zend/zend_ast.c

```diff
@@ -441,3 +444,1011 @@ ZEND_API void zend_ast_apply(zend_ast *ast, zend_ast_apply_func fn) {
 		}
 	}
 }
+
+/*
+ * Operator Precendence
+ * ====================
+ * proirity  associativity  operatiers
+ * -----------------------------------
+ *   10     left            include, include_once, eval, require, require_once
+ *   20     left            ,
+ *   30     left            or
+ *   40     left            xor
+ *   50     left            and
+ *   60     right           print
+ *   70     right           yield
+ *   80     right           =>
+ *   90     right           = += -= *= /= .= %= &= |= ^= <<= >>= **=
+ *  100     left            ? :
+ *  110     right           ??
+ *  120     left            ||
+ *  130     left            &&
+ *  140     left            |
+ *  150     left            ^
+ *  160     left            &
+ *  170     non-associative == != === !==
+ *  180     non-associative < <= > >=
+ *  190     left            << >>
+ *  200     left            + - .
+ *  210     left            * / %
+ *  220     right           !
+ *  230     non-associative instanceof
+ *  240     right           + - ++ -- ~ (type) @
+ *  250     right           **
+ *  260     left            [
+ *  270     non-associative clone new
+ */
+
+static void zend_ast_export_ex(smart_str *str, zend_ast *ast, int priority, int indent);
+
+static void zend_ast_export_str(smart_str *str, zend_string *s)
+{
+	size_t i;
+
+	for (i = 0; i < s->len; i++) {
+		unsigned char c = s->val[i];
+		if (c == '\'' || c == '\\') {
+			smart_str_appendc(str, '\\');
+			smart_str_appendc(str, c);
+		} else {
+			smart_str_appendc(str, c);
+		}
+	}
+}
+
+static void zend_ast_export_dstr(smart_str *str, zend_string *s)
+{
+	size_t i;
+
+	for (i = 0; i < s->len; i++) {
+		unsigned char c = s->val[i];
+		if (c < ' ') {
+			switch (c) {
+				case '\n':
+					smart_str_appends(str, "\\n");
+					break;
+				case '\r':
+					smart_str_appends(str, "\\r");
+					break;
+				case '\t':
+					smart_str_appends(str, "\\t");
+					break;
+				case '\f':
+					smart_str_appends(str, "\\f");
+					break;
+				case '\v':
+					smart_str_appends(str, "\\v");
+					break;
+#ifdef PHP_WIN32
+				case VK_ESCAPE:
+#else
+				case '\e':
+#endif
+					smart_str_appends(str, "\\e");
+					break;
+				default:
+					smart_str_appends(str, "\\0");
+					smart_str_appendc(str, '0' + (c / 8));
+					smart_str_appendc(str, '0' + (c % 8));
+					break;
+			}
+		} else {
+			if (c == '"' || c == '`' || c == '$' || c == '\\') {
+				smart_str_appendc(str, '\\');
+			}
+			smart_str_appendc(str, c);
+		}
+	}
+}
+
+static void zend_ast_export_indent(smart_str *str, int indent)
+{
+	while (indent > 0) {
+		smart_str_appendc(str, '\t');
+		indent--;
+	}
+}
+
+static void zend_ast_export_name(smart_str *str, zend_ast *ast, int priority, int indent)
+{
+	if (ast->kind == ZEND_AST_ZVAL) {
+		zval *zv = zend_ast_get_zval(ast);
+
+		if (Z_TYPE_P(zv) == IS_STRING) {
+			smart_str_appendl(str, Z_STRVAL_P(zv), Z_STRLEN_P(zv));
```

Comment:
Not an issue, but we have a separate `smart_str_append(str, Z_STR_P(zv))` function for this :)